### PR TITLE
Implement Map user page

### PR DESCRIPTION
**Pull request by bandinib-amzn**
*Issue #, if available:*

*Description of changes:*
Implement UI to map users and external identities to role.

**Screenshots:**
_Mockup:_
image

_Implementation:_
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/apps/configuration/app-router.tsx [hunk 1]
```diff
@@ -147,6 +150,15 @@ export function AppRouter(props: AppDependencies) {
         </Route>
         <EuiPageBody>
           <Switch>
+            <Route
+              path={buildUrl(ResourceType.roles, Action.view, ':roleName', SubAction.mapuser)}
```
Comment (tianleh) [inline public/apps/configuration/app-router.tsx:154]: I think view is used for read only page. This looks like Action.edit

File: public/apps/configuration/app-router.tsx [hunk 2]
```diff
@@ -147,6 +150,15 @@ export function AppRouter(props: AppDependencies) {
...
+              render={(match) => (
+                <RoleView
```
Comment (tianleh) [inline public/apps/configuration/app-router.tsx:156]: Open for discussion. I think editing mapped user is better to have its own component similar to RoleEdit, say RoleEditMappedUser. @yuxitang-amzn

File: public/apps/configuration/app-router.tsx (hunk 1 above)
Comment (bandinib-amzn) [reply public/apps/configuration/app-router.tsx:154]: It depends on whether we want to consider 'mapuser' as general action. According to me mapuser is not a general action. It is specific to role. But yes let's wait for other reviewer's opinion on this.

File: public/apps/configuration/app-router.tsx (hunk 2 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/app-router.tsx:156]: This sounds reasonable. Though on the UI it is attached to role view page, in the coding part we could still name it with 'edit' which looks making more sense.

File: public/apps/configuration/app-router.tsx (hunk 1 above)
Comment (yuxitang-amzn) [reply public/apps/configuration/app-router.tsx:154]: Agreed on both. So in the url can we use role/edit/roleName/mapuser?

File: public/apps/configuration/panels/role-mapping/external-identities-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  const panel = externalIdentities.map((externalIdentity, arrayIndex) => {
+    const onValueChangeHandler = (externalIdentityToUpdate: string) =>
+      updateElementInArrayHandler(setExternalIdentities, [arrayIndex, externalIdentityToUpdate]);
+
+    return (
+      <Fragment key={`externalIdentity-${arrayIndex}`}>
+        <EuiFlexGroup>
+          <EuiFlexItem grow={false}>
+            <FormRow headerText={arrayIndex === 0 ? 'External identities' : ''}>
+              <EuiFieldText
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/role-mapping/external-identities-panel.tsx:74]: I think we need to stretch to a fixed width as in mockup.

File: public/apps/configuration/panels/role-mapping/types.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { EuiComboBoxOptionOption } from '@elastic/eui';
+
+export type ComboBoxOptions = EuiComboBoxOptionOption[];
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/role-mapping/types.tsx:18]: I lifted the same type to upper folder level in recent changes:
https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/blob/67661c972465e90dc1295da9302f72c4c035046a/public/apps/configuration/types.ts#L16-L18

Let's share the same one.

File: public/apps/configuration/panels/role-view/role-view.tsx [hunk 5]
```diff
@@ -54,12 +78,106 @@ const mappedUserColumns = [
...
+}
+
 export function RoleView(props: RoleViewProps) {
   const duplicateRoleLink = buildHashUrl(ResourceType.roles, Action.duplicate, props.roleName);
 
   const [mappedUsers, setMappedUsers] = useState<MappedUsersListing[]>([]);
   const [errorFlag, setErrorFlag] = useState(false);
   const [selection, setSelection] = useState<MappedUsersListing[]>([]);
+  const [internalUsers, setInternalUsers] = useState<ComboBoxOptions>([]);
+  const [exterIdentities, setExternalIdentities] = useState<ExternalIdentityStateClass[]>([]);
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/role-view/role-view.tsx:105]: Is it intended not to be `externalIdentities`?

File: public/apps/configuration/panels/role-view/role-view.tsx [hunk 6]
```diff
@@ -54,12 +78,106 @@ const mappedUserColumns = [
...
+        addToast(createUnknownErrorToast('fetchInternalUserNames', 'load data'));
+        console.error(e);
+      }
+    };
+
+    fetchInternalUserNames();
+  }, [addToast, props.coreStart.http]);
+  const internalUserOptions = userNames.map(stringToComboBoxOption);
+
+  const updateRollMappingHandler = async () => {
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/role-view/role-view.tsx:154]: Typo Roll

File: public/apps/configuration/panels/role-view/role-view.tsx [hunk 7]
```diff
@@ -166,7 +293,13 @@ export function RoleView(props: RoleViewProps) {
     },
   ];
 
-  return (
+  const [selectedTab, setSelectedTab] = useState(tabs[0]);
+
+  const onTabClick = (tab) => {
+    setSelectedTab(tab);
+  };
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/role-view/role-view.tsx:300]: This function looks redundant.

File: public/apps/configuration/app-router.tsx (hunk 1 above)
Comment (bandinib-amzn) [reply public/apps/configuration/app-router.tsx]: Changed the url to `role/edit/roleName/mapuser`

image

File: public/apps/configuration/app-router.tsx (hunk 2 above)
Comment (bandinib-amzn) [reply public/apps/configuration/app-router.tsx]: Agreed. Moved map user part to it's own component named `RoleEditMappedUser`

File: public/apps/configuration/panels/role-mapping/external-identities-panel.tsx (hunk 3 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/role-mapping/external-identities-panel.tsx:74]: fixed.

File: public/apps/configuration/panels/role-mapping/types.tsx (hunk 4 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/role-mapping/types.tsx]: corrected.

File: public/apps/configuration/panels/role-view/role-view.tsx (hunk 5 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/role-view/role-view.tsx]: No. It was typo. Corrected.

File: public/apps/configuration/panels/role-view/role-view.tsx (hunk 6 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/role-view/role-view.tsx]: corrected.

File: public/apps/configuration/panels/role-view/role-view.tsx (hunk 7 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/role-view/role-view.tsx]: removed.

**Pull request merged**